### ISSUE TITLE
WALinuxAgent-2.2.14 did not terminate cleanly

### ISSUE THREAD

**Issue by PauloAugusto-Asos** [Azure/WALinuxAgent#828]
Hi,

Yesterday without any manual intervention, at \~20:00, the CPU of my server got stuck at a permanent 5~10% and continuous high rate of write IOPS to the O.S. drive.

I see lots of short-lived [python3] processes showing up, just to shortly disappear. In /var/log/waagent.log I'm getting heaps of these logs:

```
2017/08/02 15:09:55 [Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux-1.3.127.7] Change log file to /var/log/azure/Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux/1.3.127.7/extension.log
q2017/08/02 15:09:57.713136 INFO Event: name=Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux, op=Enable, message=Launch command succeeded: omsagent.py -enable
2017/08/02 15:09:57.745986 INFO Event: name=WALinuxAgent, op=ProcessGoalState, message=
2017/08/02 15:09:57.747899 WARNING Agent WALinuxAgent-2.2.12 failed loading error state: Expecting value: line 1 column 1 (char 0)
2017/08/02 15:09:57.750303 WARNING Agent WALinuxAgent-2.2.14 failed with exception: 'NoneType' object has no attribute 'is_sentinel'
2017/08/02 15:09:57.752121 WARNING Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/azurelinuxagent/ga/update.py", line 280, in run
    test_agent = self.get_test_agent()
  File "/usr/lib/python3/dist-packages/azurelinuxagent/ga/update.py", line 344, in get_test_agent
    agents = [agent for agent in self._load_agents() if agent.is_test]
  File "/usr/lib/python3/dist-packages/azurelinuxagent/ga/update.py", line 344, in <listcomp>
    agents = [agent for agent in self._load_agents() if agent.is_test]
  File "/usr/lib/python3/dist-packages/azurelinuxagent/ga/update.py", line 712, in is_test
    return self.error.is_sentinel and self.supported.is_supported
AttributeError: 'NoneType' object has no attribute 'is_sentinel'

2017/08/02 15:09:58.484305 WARNING Agent WALinuxAgent-2.2.14 launched with command 'python3 -u /usr/sbin/waagent -run-exthandlers' failed with return code: 1
2017/08/02 15:09:58.486457 ERROR Event: name=WALinuxAgent, op=Enable, message=Agent WALinuxAgent-2.2.14 launched with command 'python3 -u /usr/sbin/waagent -run-exthandlers' failed with return code: 1
2017/08/02 15:09:58.488186 WARNING Agent WALinuxAgent-2.2.14 launched with command 'python3 -u /usr/sbin/waagent -run-exthandlers' returned code: 1
2017/08/02 15:09:58.489377 WARNING Agent WALinuxAgent-2.2.12 failed loading error state: Expecting value: line 1 column 1 (char 0)
2017/08/02 15:09:58.493122 INFO Installed Agent WALinuxAgent-2.2.14 is the most current agent
2017/08/02 15:09:58.621322 INFO Agent WALinuxAgent-2.2.14 is running as the goal state agent
2017/08/02 15:09:58.623671 INFO Wire server endpoint:168.63.129.16
2017/08/02 15:09:58.627366 INFO Event: name=WALinuxAgent-2.2.14, op=HeartBeat, message=
2017/08/02 15:09:58.628100 INFO Start env monitor service.
2017/08/02 15:09:58.629376 INFO Configure routes
2017/08/02 15:09:58.630748 INFO Gateway:None
2017/08/02 15:09:58.632514 INFO Routes:None
2017/08/02 15:09:58.652854 INFO WALinuxAgent-2.2.14 running as process 38090
2017/08/02 15:09:58.653670 INFO WALinuxAgent-2.2.14 did not terminate cleanly
2017/08/02 15:09:58.654683 ERROR Event: name=WALinuxAgent, op=Restart, message=WALinuxAgent-2.2.14 did not terminate cleanly
2017/08/02 15:09:58.656155 INFO Wire server endpoint:168.63.129.16
2017/08/02 15:09:58.682150 WARNING Agent WALinuxAgent-2.2.12 failed loading error state: Expecting value: line 1 column 1 (char 0)
2017/08/02 15:09:58.685161 INFO Wire server endpoint:168.63.129.16
2017/08/02 15:09:58.713356 INFO [Microsoft.Azure.Extensions.LinuxAsm-2.2.5] Target handler state: uninstall
2017/08/02 15:09:58.714435 INFO [Microsoft.Azure.Extensions.LinuxAsm-2.2.5] [Uninstall] current handler state is: notinstalled
2017/08/02 15:09:58.738916 INFO [Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux-1.3.127.7] Target handler state: enabled
2017/08/02 15:09:58.740393 INFO [Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux-1.3.127.7] [Enable] current handler state is: enabled
2017/08/02 15:09:58.741185 INFO [Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux-1.3.127.7] Update settings file: 0.settings
2017/08/02 15:09:58.742255 INFO [Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux-1.3.127.7] Enable extension [omsagent.py -enable]
2017/08/02 15:09:58 OmsAgentForLinux started to handle.
2017/08/02 15:09:58 [Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux-1.3.127.7] cwd is /var/lib/waagent/Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux-1.3.127.7
2017/08/02 15:09:58 [Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux-1.3.127.7] Change log file to /var/log/azure/Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux/1.3.127.7/extension.log
```

**Comment by eamonnfaherty** [comment]
+1

**Comment by brendandixon** [comment]
@PauloAugusto-Asos The agent records with each update it downloads a small JSON file used to track health status (`error.json`). The messages above (notably "Agent WALinuxAgent-2.2.12 failed loading error state") indicate this file is either missing or corrupted. We've seen a couple of occurrences but do not have a cause identified nor can we recreate the problem.

We have, however, in pull-request #794 change the affected code such that it is more robust. These changes will be available in our next release (2.2.15).
